mbedtls/src/pk/mod.rs

```diff
@@ -960,16 +960,18 @@ impl Pk {
                 )?;
                 ecdh.calc_secret(shared, rng)
             },
-            _ => return Err(Error::PkTypeMismatch),
+            _ => return Err(Error::from(codes::PkTypeMismatch)),
         }
     }
 
     pub fn write_private_der<'buf>(&mut self, buf: &'buf mut [u8]) -> Result<Option<&'buf [u8]>> {
         match unsafe {
             pk_write_key_der(&mut self.inner, buf.as_mut_ptr(), buf.len()).into_result()
         } {
-            Err(Error::Asn1BufTooSmall) => Ok(None),
-            Err(e) => Err(e),
+            Err(e) => match e.low_level() {
+                Some(codes::Asn1BufTooSmall) => Ok(None),
+                _ => Err(e)
+            }
```

Comment:
Nested `match` can be avoided like so: (same for similar cases below and in certificate.rs and csr.rs)
```suggestion
            Err(e) => if e.low_level() == Some(codes::Asn1BufTooSmall) => Ok(None),
            Err(e) => Err(e),
```